verification/io/io-spec_definitions.gobra

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 ETH Zurich
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +gobra
+
+package io
+
+/*** All definitions/functions here are not part of the Isabelle specifications ***/
+
+
+// Establishes the traversed segment for packets which are not incremented (internal).
+ghost
+requires len(currseg.Future) > 0
+decreases
+pure func establishGuardTraversedseg(currseg IO_seg3, direction bool) IO_seg3 {
+	return let uinfo := direction ?
+		upd_uinfo(currseg.UInfo, currseg.Future[0]) :
+		currseg.UInfo in
+		IO_seg3_ {
+			AInfo: currseg.AInfo,
+			UInfo: uinfo,
+			ConsDir: currseg.ConsDir,
+			Peer: currseg.Peer,
+			Past: currseg.Past,
+			Future: currseg.Future,
+			History: currseg.History,
+		}
+}
+
+// Establishes the traversed segment for packets that are incremented (external).
+ghost
+requires len(currseg.Future) > 0
+decreases
+pure func establishGuardTraversedsegInc(currseg IO_seg3, direction bool) IO_seg3 {
+	return let uinfo := direction ?
+		upd_uinfo(currseg.UInfo, currseg.Future[0]) :
+		currseg.UInfo in
+		IO_seg3_ {
+			AInfo: currseg.AInfo,
+			UInfo: uinfo,
+			ConsDir: currseg.ConsDir,
+			Peer: currseg.Peer,
+			Past: seq[IO_HF]{currseg.Future[0]} ++ currseg.Past,
+			Future: currseg.Future[1:],
+			History: seq[IO_ahi]{currseg.Future[0].Toab()} ++ currseg.History,
+		}
+}
+
+ghost
+decreases
+pure func (seg IO_seg3) UpdateCurrSeg( info AbsInfoField) (IO_seg3) {
+	return IO_seg3_ {
+		info.AInfo,
+		info.UInfo,
+		info.ConsDir,
+		info.Peer,
+		seg.Past,
+		seg.Future,
+		seg.History,
+	}
+}
+
+ghost
+decreases
+pure func UpdateInfoField(oldPkt IO_pkt2, info AbsInfoField) (newPkt IO_pkt2) {
+	return let newCurrSeg := oldPkt.CurrSeg.UpdateCurrSeg(info) in
+		IO_Packet2{newCurrSeg, oldPkt.LeftSeg, oldPkt.MidSeg, oldPkt.RightSeg}
+}
+
+// This type simplifies the infoField, making it easier
+// to use than the IO_seg3 from the IO-spec.
+type AbsInfoField adt {
+	AbsInfoField_ {
+		AInfo IO_ainfo
+		UInfo set[IO_msgterm]
+		ConsDir bool
+		Peer bool
+	}
+}
+
+// The segment lengths of a packet are frequently used together.
+// This type combines them into a single structure to simplify
+// their specification.
+type SegLens adt {
+	SegLens_ {
+		SegLen seq[int]
```

Comment:
I would inline the entries: instead of a single field of type `seq`, I would have fields `NumSegs`, `Seg1Len`, `Seg2Len`, and `Seg3Len`. With this, you do not need to propagate any kind of properties about the length

Reply:
You may not even need `NumSegs`, considering that you always assume it is 3

Reply:
I agree. I only need `Seg1Len`, `Seg2Len` and `Seg3Len`.